src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlDateTime2.cs

```diff
@@ -0,0 +1,136 @@
+﻿using System;
+using System.Data.SqlTypes;
+using System.Runtime.InteropServices;
+using System.Xml;
+using System.Xml.Schema;
+using System.Xml.Serialization;
+using Microsoft.Data.Common;
+
+namespace Microsoft.Data.SqlTypes
+{
+    /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlDateTime2.xml' path='docs/members[@name="SqlDateTime2"]/SqlDateTime2/*' />
+    [XmlSchemaProvider("GetXsdType")]
+    [Serializable]
+    [StructLayout(LayoutKind.Sequential)]
+    public struct SqlDateTime2 : INullable, IComparable<SqlDateTime2>, IEquatable<SqlDateTime2>, IComparable, IXmlSerializable
+    {
+        private const long _minTicks = 0; // DateTime.MinValue.Ticks
+        private const long _maxTicks = 3155378975999999999; // DateTime.MaxValue.Ticks
+
+        private bool _notNull;    // false if null - has to be _notNull and not _isNull - otherwise default(SqlDateTime2) will return a SqlDateTime2 with _isNull=false and that would make XmlSerialization fail
+        private long _ticks; // Ticks representation similar to DateTime.Tics
+        
+
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlDateTime2.xml' path='docs/members[@name="SqlDateTime2"]/ctor1/*' />
+        private SqlDateTime2(bool isNull)
+        {
+            _notNull = !isNull;
+            _ticks = 0;
+        }
+
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlDateTime2.xml' path='docs/members[@name="SqlDateTime2"]/ctor2/*' />
+        public SqlDateTime2(long ticks)
+        {
+            if (ticks < _minTicks || ticks > _maxTicks)
```

Comment:
The point of this change was to improve performance.

Having DateTime.MinValue.Ticks & DateTime.MaxValue.Ticks in two constants is much faster than having to access DateTime.MinValue.Ticks. If you insist on referencing the DateTime.MinValue/MaxValue - a compromise could be to introduce two static fields instead of two constants, i.e. 

```c#
private static readonly long _minTicks = DateTime.MinValue.Tics;
private static readonly long _maxTicks = DateTime.MaxValue.Tics;
```

But I would prefer the two constants, since they would definetely be inlined when compiled - and I am not sure about the DateTime.Min/MaxValue.

Reply:
I did some investigation on this a while back when I was looking over the change and it is right that there's a significant performance difference from using the constant. At some point i was going to look at the existing datetime code to see if we could benefit from using it there.
 
Thinking about that though the value that is being fetched is a constant to the jit should see that and inline it on it's own. I may ask about why it isn't in runtime.

Reply:
The Ticks property on the DateTime struct is not a constant, its calculated. DateTime.MinValue is a static value, so that is not a constant either. So the compiler cannot really inline anything. Furthermore I think when accessing the Ticks property - you have to copy the entire structure in memory because its not readonly.

Reply:
Can we do "MinTicks" and "MaxTicks" instead as they're constants?

https://github.com/dotnet/runtime/blob/e6a5339b9231976eccd29abb1c58298e353e4e87/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L79-L80